lib/Transforms/IPO/StructFieldCacheAnalysis.cpp

```diff
@@ -36,30 +46,672 @@ class StructFieldCacheAnalysisPass : public ModulePass {
+    if (isa<Instruction>(User)){
+      auto* Inst = dyn_cast<Instruction>(User);
+      if (Inst->getOpcode() == Instruction::Load)
+        addFieldAccessNum(Inst, FieldLoc);
+      else if (Inst->getOpcode() == Instruction::Store){
+        if (U == Inst->getOperand(1))
+          addFieldAccessNum(Inst, FieldLoc);
+      }
+      else{
+        if (Inst->getOpcode() == Instruction::Call || Inst->getOpcode() == Instruction::Invoke){
+          addStats(StructFieldAccessManager::DebugStats::DS_GepPassedIntoFunc);
+        }
+        else if (Inst->getOpcode() == Instruction::BitCast){
+          addStats(StructFieldAccessManager::DebugStats::DS_GepPassedIntoBitcast);
+        }
+        else{
+          // TODO: Collect stats of this kind of access and add analysis later
+          addStats(StructFieldAccessManager::DebugStats::DS_GepUnknownUse, Inst->getOpcode());
+        }
+      }
+    }
+    else if (isa<Operator>(User)){
+      auto* Inst = dyn_cast<Operator>(U);
+      assert (Inst->getOpcode() != Instruction::Load || Inst->getOpcode() != Instruction::Store
+              || Inst->getOpcode() != Instruction::Call || Inst->getOpcode() != Instruction::Invoke);
+      if (Inst->getOpcode() == Instruction::BitCast){
+        addStats(StructFieldAccessManager::DebugStats::DS_GepPassedIntoBitcast);
+      }
+      else{
+        // TODO: Collect stats of this kind of access and add analysis later
+        addStats(StructFieldAccessManager::DebugStats::DS_GepUnknownUse, Inst->getOpcode());
+      }
+    }
+    else {
+      addStats(StructFieldAccessManager::DebugStats::DS_UserNotInstructionNorOperator);
+    }
+  }
+}
+
+void StructFieldAccessInfo::analyzeUsersOfStructValue(const Value* V)
+{
+  assert(V->getType()->isPointerTy() && V->getType()->getPointerElementType()->isStructTy());
+  for (auto *U : V->users()){
+    DEBUG_WITH_TYPE(DEBUG_TYPE_IR, dbgs() << "Analyzing user of " << *V << ": " << *U << "\n");
+    //assert(isa<Instruction>(U) || isa<Operator>(U));
+    if (isa<Instruction>(U)){
+      auto *Inst = dyn_cast<Instruction>(U);
+      if (Inst->getOpcode() != Instruction::GetElementPtr){
+        // Only support access struct through GEP for now
+        if (Inst->getOpcode() == Instruction::Call){
+          auto* F = dyn_cast<CallInst>(Inst)->getCalledFunction();
+          if (F->isDeclaration()){
+            // If a struct is passed to a function not declared in the program, we can't analyze it...
+            Eligiblity = false;
+          }
+        }
+        else if (Inst->getOpcode() == Instruction::Invoke){
+          auto* F = dyn_cast<InvokeInst>(Inst)->getCalledFunction();
+          if (F->isDeclaration()){
+            // If a struct is passed to a function not declared in the program, we can't analyze it...
+            Eligiblity = false;
+          }
+        }
+        continue;
+      }
+      addFieldAccessFromGEP(Inst);
+    }
+    else if (isa<Operator>(U)){
```

Comment:
Because I found GEP can also be an Operator (GEPOperator) that uses a struct field, not just GetElementPtrInst